libc: Add shell support

krish2718 opened the pull request:
This is useful as a placeholder for all commands common to applications. For now we start with malloc stats, useful in debugging libc HEAP issues.

krish2718 committed:
libc: Add shell support

This is useful as a placeholder for all commands common to applications.
For now we start with malloc stats, useful in debugging HEAP issues.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

fabiobaltieri commented:
Feels like it would make more sense to leave the command "app" for the app.

For the heap one specifically, how about putting it under "kernel"? It's a kernel service, there's already "kernel stacks" and "kernel threads" etc... these are from `subsys/shell/modules/kernel_service.c`. I think the heap stats very much belong in there, though you'd probably want to wrap it in something so it could use `shell_print`.

krish2718 commented:
> Feels like it would make more sense to leave the command "app" for the app.
> 
> For the heap one specifically, how about putting it under "kernel"? It's a kernel service, there's already "kernel stacks" and "kernel threads" etc... these are from `subsys/shell/modules/kernel_service.c`. I think the heap stats very much belong in there, though you'd probably want to wrap it in something so it could use `shell_print`.

The problem is there is already a `kernel heap` command when `CONFIG_SYS_HEAP_RUNTIME_STATS` is enabled, so, as a counter part had to add a `app heap`. Don't want to put this in the `kernel_service.c` as this is purely application related, but if no one has objections and we don't foresee any other commands, happy to extend `kernel_service.c`.

fabiobaltieri commented:
> The problem is there is already a `kernel heap` command when `CONFIG_SYS_HEAP_RUNTIME_STATS` is enabled, so, as a counter part had to add a `app heap`. Don't want to put this in the `kernel_service.c` as this is purely application related, but if no one has objections and we don't foresee any other commands, happy to extend `kernel_service.c`.

What do you mean by "application related"? Sorry I'm missing context here and I'm a bit confused by:
```
$ git grep malloc_stats
$
```

krish2718 commented:
> > The problem is there is already a `kernel heap` command when `CONFIG_SYS_HEAP_RUNTIME_STATS` is enabled, so, as a counter part had to add a `app heap`. Don't want to put this in the `kernel_service.c` as this is purely application related, but if no one has objections and we don't foresee any other commands, happy to extend `kernel_service.c`.
> 
> What do you mean by "application related"? Sorry I'm missing context here and I'm a bit confused by:
> 
> ```
> $ git grep malloc_stats
> $
> ```

This is for applications using `malloc` (`libc` HEAP) and doesn't include `kmalloc` (this is covered by `kernel heap` command).

fabiobaltieri commented:
> This is for applications using `malloc` (`libc` HEAP) and doesn't include `kmalloc` (this is covered by `kernel heap` command).

Ok I see what you mean now, wondering if should not be a standalone command then. Adding the C library maintainers as reviewers.

krish2718 commented:
> > This is for applications using `malloc` (`libc` HEAP) and doesn't include `kmalloc` (this is covered by `kernel heap` command).
> 
> Ok I see what you mean now, wondering if should not be a standalone command then. Adding the C library maintainers as reviewers.

yeah, we can rename `app` to something more appropriate, suggestions welcome.

cfriedt commented on subsys/shell/modules/app_service.c:
```diff
@@ -0,0 +1,39 @@
+	/* Note: This doesn't use shell_print, so, might not work with all
+	 *       backends.
+	 */
+	malloc_stats();
```
Would be kind of nice to have `mallinfo()` support, even though IIRC, it's a gnu-ism.

Actually, the fact that `malloc_stats()` does not use the `shell_print()` is a good enough reason to add `mallinfo()` support.

krish2718 replied:
Yes, in fact my initial code was based on `mallinfo2` but latest realized it's not supported and had to resort to `malloc_stats`.

cfriedt replied:
You could add it if you like. I would approve, likely also Keith

krish2718 replied:
Will take a stab, thanks.

krish2718 replied:
Just had to include `REQUIRES_FULL_LIBC` and it worked.

fabiobaltieri commented:
> yeah, we can rename `app` to something more appropriate, suggestions welcome.

Hard to say, do you already have something in mind that should go under this?

krish2718 commented:
> > yeah, we can rename `app` to something more appropriate, suggestions welcome.
> 
> Hard to say, do you already have something more in mind that should go under this?

Not currently.

fabiobaltieri commented:
> Not currently.

Ok then I'm thinking it may make sense to either let it live as a top level command for now, maybe under `lib/libc`, and then we could regroup down the road if and when we get more and having a group becomes more natural. @cfriedt what do you think?

krish2718 commented:
> > Not currently.
> 
> Ok then I'm thinking it may make sense to either let it live as a top level command for now, maybe under `lib/libc`, and then we could regroup down the road if and when we get more and having a group becomes more natural. @cfriedt what do you think?

Moved to `libc`, now the command is `libc heap`.

jakub-uC commented on lib/libc/shell.c:
```diff
@@ -0,0 +1,38 @@
+	ARG_UNUSED(argc);
+	ARG_UNUSED(argv);
+
+	shell_fprintf(sh, SHELL_NORMAL, "Heap size: %d bytes\n", mi.arena);
```
This is fine but `shell_print` macro looks more nifty ;). You will not need `SHELL_NORMAL` and `\n`.

cfriedt reviewed (dismissed):
Given that virtually every Zephyr app includes the libc, to me it seems weird that there would be a "libc shell".

I think this should go under `shell/modules/app_service.c` in order to differentiate it from the `kernel heap` command.

e.g. `app heap` vs `kernel heap` (I can't recall if that was the exact command name)

krish2718 commented:
> Given that virtually every Zephyr app includes the libc, to me it seems weird that there would be a "libc shell".
> 
> I think this should go under `shell/modules/app_service.c` in order to differentiate it from the `kernel heap` command.
> 
> e.g. `app heap` vs `kernel heap` (I can't recall if that was the exact command name)

That was the original change but based on comment [from](https://github.com/zephyrproject-rtos/zephyr/pull/60269#issuecomment-1633903429) moved to libc, I am okay either way, but if we can agree on one then I can update this PR.

fabiobaltieri commented:
> Given that virtually every Zephyr app includes the libc, to me it seems weird that there would be a "libc shell".
> 
> I think this should go under `shell/modules/app_service.c` in order to differentiate it from the `kernel heap` command.
> 
> e.g. `app heap` vs `kernel heap` (I can't recall if that was the exact command name)

Yeah I was suggesting to not use the category `app` and leave it for the app instead. Were would you put it @cfriedt? I was thinking a `libc` command would make sense if we had more of those, otherwise I would just leave it at top level for now. Or under `kernel` with a different name? I don't have a strong opinion about it, it just felt weird to me to use `app` in the system. Mmh... "system"...

nashif reviewed (changes requested):
> This is useful as a placeholder for all commands common to applications.

Hmm,  not sure what all those command are, what do you have in mind?

libc shell sounds weird, what you are doing here is  showing heap usage, this is not really a libc shell, this is a memory/heap stat command, just put is along side other commands we already have or have something new for memory, but creating a "libc shell" for the sake of just getting heap info is an overkill.

cfriedt commented:
Hmm... yeah, I get that app might not be ideal. Is "system" too vague?

fabiobaltieri commented:
> Hmm... yeah, I get that app might not be ideal. Is "system" too vague?

How about staying under kernel but with a different name? `kernel libc_heap`?

krish2718 commented:
> > Hmm... yeah, I get that app might not be ideal. Is "system" too vague?
> 
> How about staying under kernel but with a different name? `kernel libc_heap`? It's not really a kernel service though :-\
> 
> Update: how about making `kernel heap` show both the Zephyr heap and the libc one?

Not ideal esp. if CONFIG_USERSPACE is set, we need a way to clearly convey that this is for the applications and separate from the kernel. `system` also doesn't fit quite right, `app` as a shell module seems the closest fit IMHO.

krish2718 committed:
libc: common: Add support for malloc stats

If malloc is enabled in common libc, add support for malloc stats using
the mallinfo API.

Signed-off-by: Chaitanya Tata <Chaitanya.Tata@nordicsemi.no>

cfriedt reviewed (changes requested):
I recommend holding off until this is resolved

https://github.com/zephyrproject-rtos/zephyr/pull/61112

cfriedt commented on lib/libc/common/source/stdlib/malloc.c:
```diff
@@ -251,6 +275,15 @@ void *realloc(void *ptr, size_t size)
 	return malloc(size);
 }
+
+struct mallinfo2 mallinfo2(void)
```
There is no need to add this stub for mallinfo if malloc heap size is zero.

The main reason the malloc stub exists is because malloc is part of ISO C and apps still need to be able to link

cfriedt commented on lib/libc/minimal/include/stdlib.h:
```diff
@@ -16,13 +16,27 @@
 extern "C" {
 #endif
 
+struct mallinfo2 {
```
Can this be put into `mallinfo.h` instead in the common libc directory?

krish2718 replied:
I wanted to put this in `malloc.h` as per standard, but I see that malloc is implemented as part of `stdlib`.

cfriedt replied:
`malloc.h` is correct (sorry, was mistaken above due to lack of ☕). 

Regardless of where the _declaration_ of `malloc()` is (i.e. in which `.h` file), the declaration for `mallinfo()` should go into `malloc.h`.

The _implementation_ of `malloc()` is in `lib/libc/common/source/stdlib/malloc.c`, so please ensure that you also `#include <mallinfo.h>` in that file too.

cfriedt replied:
@krish2718 - can you please address this comment?

cfriedt commented on lib/libc/minimal/include/stdlib.h:
```diff
@@ -16,13 +16,27 @@
 int atoi(const char *s);
 
 void *malloc(size_t size);
+struct mallinfo2 mallinfo2(void);
```
Same - please put this into mallinfo.h

See https://man7.org/linux/man-pages/man3/mallinfo.3.html

krish2718 replied:
I wanted to put this in `malloc.h` as per standard, but I see that malloc is implemented as part of `stdlib`.

cfriedt replied:
Again, regardless of where `malloc()` is _declared_, all declarations related to `mallinfo()` should go into `malloc.h`, per the manual page.

That's a kind of fundamental thing. Otherwise, if someone were reading the man page about how to use `mallinfo2()`, and they used `#include <malloc.h>`, and then ran their build job, there would be no function prototype declared and the compiler would fail.

```cpp
#include <stdint.h>

/* no such file or directory? */
#include <malloc.h>

int main() {
  /* undefined reference to mallinfo2 */
  return (uintptr_t) mallinfo2;
}
```

cfriedt replied:
@krish2718 - can you please address this comment?

cfriedt commented:
Please also add a testcase in tests/lib/c_lib

krish2718 commented:
Just revisiting this now, the pain point to get all tests is mallinfo vs mallinfo2 availability, and AFAIK, there is no common define that can be used across libc implementations? below works for `glibc` but not for others, it just uses the `__GLIBC__` define from the cross compiler's files in build system. Is there a common macro that I can use reliably? or some kind of `has_mallinfo/has_mallinfo2` ? 

```
#if __GLIBC__ < 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ < 33)
struct mallinfo mallinfo(void)
#else
struct mallinfo2 mallinfo2(void)
#endif /* __GLIBC__ < 2 && (__GLIBC__ == 2 && __GLIBC_MINOR__ < 33) */
```

cfriedt commented:
Is there a common macro that I can use reliably? or some kind of `has_mallinfo/has_mallinfo2` ? 
> 
> ```
> #if __GLIBC__ < 2 || (__GLIBC__ == 2 && __GLIBC_MINOR__ < 33)
> struct mallinfo mallinfo(void)
> #else
> struct mallinfo2 mallinfo2(void)
> #endif /* __GLIBC__ < 2 && (__GLIBC__ == 2 && __GLIBC_MINOR__ < 33) */
> ```

I would venture to say that the `mallinfo2` variant is the only one that we will need. If you google "glibc release dates" you'll see that glibc 1.x was last released in 1995.

However, would it be ok to simplify it to just be called `struct mallinfo`?

cfriedt reviewed (changes requested):
Please see the changes requested under "Files changed"
